### PR TITLE
Improve training display with tps, GPU Peak Mem %, cudaMallocCache retries, and do it in color for pizzaz

### PR DESCRIPTION
**Pull request by lessw2020**
This PR addresses a tune pain point that AWS has complained about for some time...specifically, they want to be able to quickly see a runs tps, gpu peak mem %, and cudaMallocCache retries directly on the console so they can quickly adjust their batch size and optimize throughput. 
Some of this info exists already but it's not shown on the console and is thus available only by pulling up tensorboard or similar, vs seeing it direclty on the console is much faster for optimizing a run. 

Thus:
1 - adds a GPUMemoryMonitor class that monitors mem stats and cudaMalloc retries and reports those
2 - displays GPU total memory at start
3 - integrates these into the tqdm display along with a bit of slight rounding to make things fit nicely.  Note this only changes display re: rounding, does not affect the underlying metrics sent to tensorboard, etc. 
4 - with those changes you get this:
Screenshot 2024-08-16 at 5 48 25 PM

and finally to add some visual excitement..
5 - adds color class and integrates color into the console display resulting in this:
Screenshot 2024-08-16 at 6 05 50 PM

I'm posting this PR now so that AWS can use these changes.  There's probably some discussion to be had as I see there is a couple simpler memory apis already built and maybe this should be consolidated etc. 
But happy to. discuss - in the interim will send this PR over for immediate use. 

#### Context
What is the purpose of this PR? Is it to
- [ X] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](../CONTRIBUTING.md) for some guidance on contributing.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](../tests/torchtune) for any new functionality
- [ ] update [docstrings](../docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [ ] I have added an example to docs or docstrings;

### CODE REVIEW

File: recipes/lora_finetune_single_device.py [hunk 1]
```diff
@@ -596,14 +611,20 @@ def train(self) -> None:
                         self.global_step += 1
 
                         loss_to_log = running_loss.item()
+                        rounded_loss = round(loss_to_log, 5)
                         pbar.update(1)
                         pbar.set_description(
-                            f"{curr_epoch + 1}|{self.global_step}|Loss: {loss_to_log}"
+                            f"{curr_epoch + 1}|{self.global_step}|{self.color.green}Loss: {rounded_loss} "
+                            + f"{self.color.white}| {self.color.blue}TPS: {self.tps} {self.color.white}|"
+                            + f"{self.color.yellow} Mem: {self.mem_pct:.2f}% "
```
[the remark this answers is not in the document]
Comment (felipemello1) [reply recipes/lora_finetune_single_device.py:619]: We would be abstracting some of these lines so the recipe is cleaner and more flexible. But in torchtune we like to have things exposed. So I am not sure.

File: recipes/lora_finetune_single_device.py [hunk 2]
```diff
@@ -596,14 +611,20 @@ def train(self) -> None:
...
+                            + f"{self.color.white}| {self.color.red} Retries: {self.mem_cache_retries}{self.color.reset}"
                         )
 
                         # Log per-step metrics
                         if self.global_step % self._log_every_n_steps == 0:
                             time_per_step = time.perf_counter() - t0
+                            self.tps = round(num_tokens / time_per_step, 3)
```
Comment (felipemello1) [inline recipes/lora_finetune_single_device.py:626]: this is calculated in line 631. We should probably have only one reference for it. Maybe update line 631?

nit: toks_per_sec should be more friendly, if you dont think its too long

File: recipes/lora_finetune_single_device.py [hunk 3]
```diff
@@ -219,6 +219,14 @@ def setup(self, cfg: DictConfig) -> None:
             ),
         )
 
+        self.gpu_monitor = None
+        if self._device.type == "cuda":
+            self.gpu_monitor = utils.memory.build_gpu_memory_monitor()
+        self.tps = 0
+        self.mem_pct = 0
+        self.mem_cache_retries = 0
```
Comment (felipemello1) [inline recipes/lora_finetune_single_device.py:227]: I am not sure if we need these variables to have "self.". Maybe we can just initialized them in train() instead? They are not used anywhere else, as far as I can tell

**Review by felipemello1 — COMMENTED** [review]
I love the display. Its definitely helpful! I left a few comments.

File: torchtune/utils/memory.py [hunk 4]
```diff
@@ -240,3 +241,79 @@ def log_memory_stats(stats: Dict[str, float]) -> None:
...
+        max_active = cuda_info["active_bytes.all.peak"]
+        max_active_gib = self._to_gib(max_active)
+        max_active_pct = self._to_pct(max_active)
+
+        max_reserved = cuda_info["reserved_bytes.all.peak"]
+        max_reserved_gib = self._to_gib(max_reserved)
+        max_reserved_pct = self._to_pct(max_reserved)
+
+        num_retries = cuda_info["num_alloc_retries"]
+        num_ooms = cuda_info["num_ooms"]
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply torchtune/utils/memory.py:294]: Hi @felipemello1  - 
1 - how can num_retries be used:  The num retries means the number of times cudaCachingAllocator did not have enough memory to dispense for a request.  Therefore, it drops all memory and then re-requests it via cudaMalloc to see if the lack of available memory is due to fragmentation...often it is, so training can continue but it is a time hit due to drop and restart all caching. 
Generally, even a single retry is a perf hit, but certainly only seeing 1 vs 5 in an iter is the degree of perf hit.  Thus, it just provides more data as to how badly/often you are taking perf hits.

[the remark this answers is not in the document]
Comment (lessw2020) [reply torchtune/utils/memory.py:294]: 2 - can num_oom be >0 and the run not fail?  Yes, usually not in standard training but apparently in notebooks OOMs can happen and continue.  
This metric is generally not that useful since we are inherently running in non-interactive training/tuning, so that's why it's not displayed atm.  We can just remove this from tracking though.

**Review by kartikayk — COMMENTED** [review]
@lessw2020 thanks so much for making these changes! I think our logging to console can definitely be better and these metrics generally make a lot of sense to me.

I'll let you answer some of @felipemello1 questions and also tag @ebsmothers who's done a ton of work on logging in the past.

My primary suggestion would be around the design of this component and how we want to expose this in each recipe. It seems like the current method has a ton of functionality which will need to be copied over in each recipe and this runs the risk of fragmenting the logic as more changes are made and of course bugs resulting from duplication.

I'd suggest designing a stateful class that we can initialize during setup of the recipe, which tracks all of the information needed as part of its state and exposes some APIs which can return the relevant information to the ```train``` method. I think it's ok to expect the recipe to do the logging of the actual metrics. I think this class can also run some aggregates (eg: max, avg etc) which we can log at the end of training to give some more insights. A few benefits of this approach:
- The logic is consolidated in a single class and so all issues like distributed vs single GPU and interaction with other tooling can be addresses in one place
- Any updates to logic can be automatically shared across recipes rather than copy pasting this again and again

Does this make sense?

**Comment by lessw2020** [comment]
> @lessw2020 thanks so much for making these changes! I think our logging to console can definitely be better and these metrics generally make a lot of sense to me.
Thanks!

> I'll let you answer some of @felipemello1 questions and also tag @ebsmothers who's done a ton of work on logging in the past.
Will do.

> My primary suggestion would be around the design of this component and how we want to expose this in each recipe. It seems like the current method has a ton of functionality which will need to be copied over in each recipe and this runs the risk of fragmenting the logic as more changes are made and of course bugs resulting from duplication.

100% agree - that's exactly what I meant about the need to discuss better integration overall...this initial PR was to show it working and get something to AWS asap for their immediate use case (single gpu lora), and then work on a better integration assuming you guys liked the concept overall. 

> I'd suggest designing a stateful class that we can initialize during setup of the recipe, which tracks all of the information needed as part of its state and exposes some APIs which can return the relevant information to the `train` method. I think it's ok to expect the recipe to do the logging of the actual metrics. I think this class can also run some aggregates (eg: max, avg etc) which we can log at the end of training to give some more insights. A few benefits of this approach:
Agree - this is very similar to how we implemented things in titan btw. 
 
> * The logic is consolidated in a single class and so all issues like distributed vs single GPU and interaction with other tooling can be addresses in one place
> * Any updates to logic can be automatically shared across recipes rather than copy pasting this again and again
>
> Does this make sense?

Yes seems ideal from a design standpoint.

File: recipes/lora_finetune_single_device.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lessw2020) [reply recipes/lora_finetune_single_device.py:619]: I like the cleanness of the above. I think I'll update to it unless there are other design objections.

File: recipes/lora_finetune_single_device.py (hunk 2 above)
Comment (lessw2020) [reply recipes/lora_finetune_single_device.py:626]: yes on both points.

File: torchtune/utils/memory.py [hunk 5]
```diff
@@ -240,3 +241,79 @@ def log_memory_stats(stats: Dict[str, float]) -> None:
...
+        "max_reserved_pct",
+        "num_alloc_retries",
+        "num_ooms",
+    ],
+)
+
+
+class GPUMemoryMonitor:
+    def __init__(self, device: str = "cuda:0"):
+        self.device = torch.device(device)  # device object
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply torchtune/utils/memory.py:262]: I can check how we did it in titan - I think it will work but need to check.

File: torchtune/utils/memory.py [hunk 6]
```diff
@@ -240,3 +241,79 @@ def log_memory_stats(stats: Dict[str, float]) -> None:
...
+
+    def _to_pct(self, memory):
+        return 100 * memory / self.device_capacity
+
+    def get_peak_stats(self):
+        cuda_info = torch.cuda.memory_stats(self.device)
+
+        max_active = cuda_info["active_bytes.all.peak"]
+        max_active_gib = self._to_gib(max_active)
+        max_active_pct = self._to_pct(max_active)
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply torchtune/utils/memory.py:287]: ah here I think the % is important.  We recommend people shoot for 85-90% utilization while ensuring no retries as the optimal path to max throughput.
If we show 72/95...now the user has to manually convert this to actionable info. 
If we could fit both (75% | 72/95) then that's fine, but if picking or choosing then I don't see gain from making the user do the % math each time.

File: torchtune/utils/memory.py [hunk 7]
```diff
@@ -240,3 +241,79 @@ def log_memory_stats(stats: Dict[str, float]) -> None:
...
+
+        num_retries = cuda_info["num_alloc_retries"]
+        num_ooms = cuda_info["num_ooms"]
+
+        if num_retries > 0:
+            _log.warning(f"{num_retries} CUDA memory allocation retries.")
+        if num_ooms > 0:
+            _log.warning(f"{num_ooms} CUDA OOM errors thrown.")
+
+        return max_reserved_pct, num_retries
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply torchtune/utils/memory.py:301]: both are useful, but between the two reserved is what determines both retries and OOM...active is more useful for people doing dev work on a feature to monitor the net memory, or as you noted, they are constantly emptying the cache though that itself can cause perf issues (the caching allocator hangs onto it for a reason). 
We displayed both for a bit in titan but ultimately just went with reserved given space constraints.

File: torchtune/utils/memory.py [hunk 8]
```diff
@@ -240,3 +241,79 @@ def log_memory_stats(stats: Dict[str, float]) -> None:
...
+        # )# GPUMemStats(
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply torchtune/utils/memory.py:302]: Probably don't need it - for sure we only need one memory monitor class. 
I can update to centralize and remove.

File: torchtune/utils/memory.py [hunk 9]
```diff
@@ -240,3 +241,79 @@ def log_memory_stats(stats: Dict[str, float]) -> None:
...
+        # max_active_gib,
+        # max_active_pct,
+        # max_reserved_gib,
+        # num_ooms,
+
+    def reset_peak_stats(self):
+        torch.cuda.reset_peak_memory_stats()
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply torchtune/utils/memory.py:309]: we should test, but intrinsically torch profiler does not monitor memory - it monitors compute and comms. 
also this reset is normally only fired at the start of training, and then usually not touched.
